### PR TITLE
Add ARCHITECTURE.rst and module docstrings

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This PR adds high level documentation for Connexion
- An `ARCHITECTURE.rst` file with a diagram of the high level architecture and a short explanation of the main components
- Module docstrings for each module

Since documentation on the architecture and internals was only sparsely available, I had to build my understanding mostly by going through the code. This wasn't always straightforward so let me know if you think I misinterpreted / misrepresented something, or if some descriptions aren't clear.

### CODE REVIEW

File: connexion/apis/__init__.py [hunk 1]
```diff
@@ -1 +1,16 @@
+"""
+This module defines an Connexion APIs. A connexion API takes in an OpenAPI specification and
```
Comment (Ruwann) [inline connexion/apis/__init__.py:2]: typo: "an Connexion APIs."

File: connexion/apps/flask_app.py [hunk 2]
```diff
@@ -1,3 +1,7 @@
+"""
+This module defines a FlaskApp, a Connexion application to wrap an AioHttp application.
```
Comment (Ruwann) [inline connexion/apps/flask_app.py:2]: "to wrap a Flask application"

File: connexion/decorators/response.py [hunk 3]
```diff
@@ -1,4 +1,7 @@
-# Decorators to change the return type of endpoints
+"""
+This module defines an view function decorator to validate its responses.
```
Comment (Ruwann) [inline connexion/decorators/response.py:2]: "an view function" --> "a view function"

File: ARCHITECTURE.rst [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+Connexion requests and responses
+--------------------------------
+
+Connexion defines a request and response interface for internal use. The outermost decorator of
+the operation casts framework specific requests to ``ConnexionRequests`` and ``ConnexionResponses``
+to framework specific responses.
+
+.. _Flask blueprint: https://flask.palletsprojects.com/en/2.0.x/blueprints/
+.. _OpenAPI operation: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#operation-object
```
Comment (Ruwann) [inline ARCHITECTURE.rst:73]: Perhaps provide a link to both swagger 2 and OpenAPI 3 operation objects to make it clear both versions are supported?

https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#operationObject

**Review by Ruwann — COMMENTED** [review]
AFAICT, the architecture seems logical to me. Haven't been able to check everything, so I'll continue when I have tried it out.

Minor comment on the diagram is that the `1:n` are not super clear, I would perhaps use a similar notation to UML diagrams by indicating the cardinality at both ends of the relationship.

**Comment by RobbeSneyders** [comment]
Closing as this has been merged upstream.

**Pull request closed without merging**